data/base/sub-endpoint.php

```diff
@@ -13,27 +11,78 @@ abstract class SubEndpoint extends Endpoint {
 	/**
 	 * @var string
 	 */
-	protected $parent_route = '';
+	protected $route = '';
 
-	public function __construct( $parent_route, $parent_endpoint ) {
+	public function __construct( $parent_endpoint, $route = '' ) {
+		$this->route = $route;
 		$this->parent_endpoint = $parent_endpoint;
-		$this->parent_route = $parent_route;
 
 		parent::__construct( $this->parent_endpoint->controller );
 	}
 
+	public function get_format() {
+		return $this->get_base_route();
+	}
+
 	/**
 	 * Get parent route.
 	 *
-	 * @return \Elementor\Data\Base\Endpoint
+	 * @return \Elementor\Data\Base\Endpoint|\Elementor\Data\Base\SubEndpoint
 	 */
 	public function get_parent() {
 		return $this->parent_endpoint;
 	}
 
 	public function get_base_route() {
-		$controller_name = $this->controller->get_name();
+		$name = $this->get_name();
+		$parent = $this->get_parent();
+		$is_parent_sub_endpoint = $parent instanceof SubEndpoint;
+
+		$route = $this->route;
+
+		if ( $route ) {
+			$route .= '/';
+		}
+
+		// Parent sub endpoint
+		if ( $is_parent_sub_endpoint ) {
+			return $parent->get_base_route() . '/' . $route . $name;
+		}
+
+		// Parent sub index endpoint.
+		if ( $this->controller instanceof SubController && $parent instanceof Endpoint\Index\SubController ) {
```

Comment:
The base class shouldn't know about parents logic.

Reply:
Kinda handled: https://github.com/elementor/elementor/pull/12820/commits/b4cdcd462691d96b8f0b3424c8e636315098bab2
is `$this->controller instanceof SubController` that fine?